mlir/include/mlir/Interfaces/InferTypeOpInterface.td

```diff
@@ -142,6 +142,9 @@ def InferShapedTypeOpInterface : OpInterface<"InferShapedTypeOpInterface"> {
       Unknown (e.g., unranked) shape and nullptrs for element type and attribute
       may be returned by this function while returning success. E.g., partial
       population of components is not error condition.
+
+      Be aware that this method is supposed to be called with valid arguments,
+      e.g., operands are verified, or it may result in an undefined behavior.
```

Comment:
This is the opposite of my understanding of what @jpienaar wrote on Discord. This method can be called by the verifier and so can't assume anything about the operands.

Reply:
The reason I originally suggested on discord to copy this sentence over is that both `InferTypeOpInterface` and `InferShapedTypeOpInterface` have the same phrase in the `description`, saying: "[...] could be used during op construction, verification or type inference", but only one (`inferTypeOpInterface`) has the phrase saying "the operands need to be verified". I understood @jpienaar's explanation as meaning that since both methods could be used under the same circumstances (i.e. potentially verifier), it would be consistent for them to have the same requirements about operands. Did I misunderstand this?

I was trying to get a better understanding of how these methods are used but couldn't find anything in tree, could you point me to an example? I guess I don't understand why the possibility of using the methods in the verifier means that they cannot assume the operands are valid? The verifier could check the operands and attributes, and if it hasn't errored out by then, it could call `inferReturnType/Components`.

Reply:
It's an ordering question: the verifier one can set different orderings, and AFAIK verification of number of operands etc occur before trait verification (but please check me on this) so that it does handle many cases in the verifier. But its also called in the builder, and in the builder those have often not been verified yet (that could be improved with the OpAdaptor verifier).

These do assume the IR/operands are valid, but that's also true for almost every function and interface (calling on invalid IR is rather unbounded set). Now valid here has an asterisk, valid doesn't for example mean the types are correct, just that it has the correct number of arguments say and could fail gracefully here. Rank verification and the like also happen inside this quite oftn (as verifying the input rank is often done along with computing the result). Ideally we'd have more of these declaratively and generate these functions more so that the ordering can be taken care off by the pre- and post-conditions.

Reply:
The Builder API is a different story, we can expect a builder call from a pass to not be incorrect, that's a general approach in MLIR (it's not just about Builder, but any kind of transformation, and also Builder is not an API that is allowed to fail, so the client must provide correct arguments).

The "valid IR" thing is really about the verifier to me, either we can guarantee things that are verified, or we should treat the input as "unsafe" (which is the case here).

Reply:
> or we should treat the input as "unsafe" (which is the case here)

Makes sense. I understand now why we potentially have to tolerate unverified operands here, but I still don't fully understand the difference between `inferReturnTypes` and `inferReturnTypeComponents` regarding this.

> It's an ordering question: the verifier one can set different orderings, and AFAIK verification of number of operands etc occur before trait verification (but please check me on this) so that it does handle many cases in the verifier.

Would it make sense to point out in the documentation for both methods, that they need to be able to tolerate "unsafe" (i.e. not fully verified) inputs, although the "degree of unsafeness" might vary depending at which stage of the verifier they are called?

Reply:
Yes I would document this consistently.